common/workunit/workflow.cpp

```diff
@@ -1159,21 +1144,29 @@ void WorkflowMachine::failDependentSuccessors(CCloneWorkflowItem &item)
         }
     }
 }
-bool WorkflowMachine::activateFailureContingency(CCloneWorkflowItem & item, bool isAborting)
+bool WorkflowMachine::activateFailureContingency(CCloneWorkflowItem & item)
 {
     unsigned failureWfid = item.queryFailure();
     if(failureWfid)
     {
         startContingency();
         CCloneWorkflowItem & failureActivator = queryWorkflowItem(failureWfid);
-        if(isAborting)
-            failureActivator.setAbortFlag(true);
         failureActivator.setContingencyWithin(item.queryWfid());
         processLogicalSuccessors(failureActivator);
         return true;
     }
     return false;
 }
+void WorkflowMachine::checkAbort(CCloneWorkflowItem & item)
+{
+    if(item.queryContingencyWithin())
+        return;
+    if(!abort)
+    {
+        runtimeError.set(item.queryException());
```

Comment:
does this need to be protected by a critical section - in case two threads set it at the same time?  (Could also use an atomic exchange if you want to avoid a critical section.)